File: 5.最长回文子串.java
/*
 * @lc app=leetcode.cn id=5 lang=java
 *
 * [5] 最长回文子串
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        int len;
        String res=String.valueOf(s.charAt(0));
        int max_len=1;
        for(int i=0;i<s.length();++i){
            len=1;
            for(int j=1;j<Math.min(i,s.length()-i-1)+1;++j){
                if(s.charAt(i+j)==s.charAt(i-j)){
                    len+=2;
                    if(len>max_len){
                        max_len=len;
                        res=s.substring(i-j,i+j+1);
                    }
                }
                else{
                    break;
                }
            }       
            len=0;
            for(int j=0;j<Math.min(i,s.length()-i-2)+1;++j){
                if(s.charAt(i+j+1)==s.charAt(i-j)){
                    len+=2;
                    if(len>max_len){
                        max_len=len;
                        res=s.substring(i-j,i+j+2);
                    }
                }
                else{
                    break;
                }
            }
        }
        return res;
    }
}
// @lc code=end

File: 5.最长回文子串.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] 最长回文子串
 */

// @lc code=start
#include<vector>
#include<iostream>
#include<string>
class Solution {
public:
    string longestPalindrome(string s) {
        int len=s.size();
        if(len<2){
            return s;
        }
        int maxlen=0;
        string res;
        int dp[len][len];
        for (int i = 0; i < len;++i){
            memset(dp[i],1,len);
        }
        for(int left=len-1;left>=0;--left){
            for(int right=left;right<len;++right){
                if(s[left]==s[right]){
                    if(right-left>1){
                        dp[left][right]=dp[left+1][right-1];
                    }
                    else{
                        dp[left][right]=true;
                    }
                }
                else{
                    dp[left][right]=0;
                }
                if((right-left+1>maxlen)&&dp[left][right]){
                    maxlen=right-left+1;
                    res=s.substr(left,maxlen);
                }
            }
        }
        return res;
    }
};
// @lc code=end

